File: client/src/app/services/account.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable, tap } from 'rxjs';

// Requests
export interface UserCredentials {
  username: string;
  password: string;
}
export interface OpenAccountRequest {
  username: string;
  password: string;
  email: string;
  name: string;
  surname: string;
  mobile: string;
}

// Responses
export interface TokenResponse {
  token: string;
}

export interface ErrorResponse {
  error: string;
}

export interface AccountResponse {
  id: number;
  name: string;
  surname: string;
  username: string;
  email: string;
  mobile: string;
  role: 'ADMIN' | 'USER';
  updateDate: string;
  createDate: string;
}

export function getPossibleRoles(): AccountResponse['role'][] {
  return ['ADMIN', 'USER'];
}

export interface GetAccountsRequest {
  page: number;
  size: number;
  sortBy?: string;
  sortOrder?: string;
}

export interface AccountsPaginatedResponse {
  content: AccountResponse[];
  currentPage: number;
  totalPages: number;
  totalItems: number;
}


@Injectable({
  providedIn: 'root',
})
export class AccountService {
  private readonly tokenKey = 'auth_token';
  private readonly baseUrl = '/api/account';

  constructor(private http: HttpClient) {}

  /**
   * Effettua il login dell'utente e restituisce il token JWT.
   * @param credentials Credenziali dell'utente (username e password).
   * @returns Un Observable contenente il TokenResponse o un errore.
   */
  login(credentials: UserCredentials): Observable<TokenResponse> {
    return this.http.post<TokenResponse>(`${this.baseUrl}/login`, credentials);
  }

  /**
   * Verifica la validità di un token JWT.
   * @param token Il token JWT da verificare.
   * @returns Un Observable che emette `true` se il token è valido, altrimenti `false`.
   */
  checkTokenValidity(token: TokenResponse): Observable<boolean> {
    return this.http.post<boolean>(`${this.baseUrl}/check-token`, token);
  }

  
  openAccountUser(accountRequest: OpenAccountRequest): Observable<void> {
    return this.http.post<void>(`${this.baseUrl}/user-account`, accountRequest).pipe(
      tap({
        next: (response: any) => {
          console.log('User account created successfully', response);
        },
        error: (error: any) => {
          console.error('Error creating user account', error);
        }
      })
    );
  }
  

  getAccount(id: number): Observable<AccountResponse> {
    const headers = new HttpHeaders({ 
      'Content-Type': 'application/json',
      'Use-Auth': 'true' // tells interceptor to include bearer token
    });

    return this.http.get<AccountResponse>(`${this.baseUrl}/${id}`, { headers });
  }

  getAccounts(request: GetAccountsRequest): Observable<AccountsPaginatedResponse> {
    const headers = new HttpHeaders({ 
      'Content-Type': 'application/json',
      'Use-Auth': 'true' // tells interceptor to include bearer token
    });
    return this.http.post<AccountsPaginatedResponse>(
      `${this.baseUrl}/accounts`,
      request,
      { headers }
    );
  }

  promoteAccount(id: number): Observable<any> {
    const headers = new HttpHeaders({ 
      'Content-Type': 'application/json',
      'Use-Auth': 'true' // tells interceptor to include bearer token
    });
    return this.http.patch<void>(`${this.baseUrl}/${id}/promote`, null, {headers});
  }

  demoteAccount(id: number): Observable<any> {
    const headers = new HttpHeaders({ 
      'Content-Type': 'application/json',
      'Use-Auth': 'true' // tells interceptor to include bearer token
    });
    return this.http.patch<void>(`${this.baseUrl}/${id}/demote`, null, {headers});
  }

 
}

File: account-service/src/main/java/it/disim/univaq/sose/account_service/service/AccountService.java
package it.disim.univaq.sose.account_service.service;


import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.headers.Header;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.parameters.RequestBody;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import it.disim.univaq.sose.account_service.domain.dto.*;
import jakarta.ws.rs.*;
import jakarta.ws.rs.core.Response;

/**
 * The AccountService interface defines the service methods for the account service.
 */
@Path("/api/account")
public interface AccountService {
    /**
     * Authenticate user and return JWT
     *
     * @param userCredentials
     * @return
     */
    @Operation(
            operationId = "login",
            description = "Authenticate user and return JWT",
            responses = {
                    @ApiResponse(responseCode = "200", description = "Authentication successful",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = TokenResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = TokenResponse.class))
                            }),
                    @ApiResponse(responseCode = "401", description = "Authentication failed",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                    })
            }
    )
    @POST
    @Path("/login")
    @Consumes({"application/json", "application/xml"})
    @Produces({"application/json", "application/xml"})
    Response login(@RequestBody(description = "Login", required = true, content = {@Content(mediaType = "application/json", schema = @Schema(implementation = UserCredentials.class)), @Content(mediaType = "application/xml", schema = @Schema(implementation = UserCredentials.class))}) UserCredentials userCredentials);

    /**
     * Check JWT
     *
     * @param token
     * @return
     */
    @Operation(operationId = "checkTokenResponse", description = "Check JWT",
            responses = {
                @ApiResponse(responseCode = "200", description = "Check successful",
                        content = {
                            @Content(mediaType = "application/json", schema = @Schema(implementation = Boolean.class)),
                            @Content(mediaType = "application/xml", schema = @Schema(implementation = Boolean.class))
                        }
                )
            }
    )
    @POST
    @Path("/check-token")
    @Consumes({"application/json", "application/xml"})
    @Produces({"application/json", "application/xml"})
    Boolean checkTokenResponse(@RequestBody(description = "Token", required = true, content = {@Content(mediaType = "application/json", schema = @Schema(implementation = TokenResponse.class)), @Content(mediaType = "application/xml", schema = @Schema(implementation = TokenResponse.class))}) TokenResponse token);

    /**
     * Open Admin Account
     *
     * @param openAccountRequest
     * @return
     */
    @Operation(operationId = "openAccountAdmin", description = "openAccountAdmin",
            responses = {
                @ApiResponse(responseCode = "201", description = "Save Admin Account",
                        content = {
                                @Content(mediaType = "application/json", schema = @Schema(implementation = AccountResponse.class)),
                                @Content(mediaType = "application/xml", schema = @Schema(implementation =  AccountResponse.class))
                        },
                        headers = {@Header(name = "Location", description = "URL of the created resource", schema = @Schema(type = "string"))}
                ),
                @ApiResponse(responseCode = "500", description = "Account whit this Id not found",
                        content = {
                            @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                            @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                        }
               )
            }
    )
    @POST
    @Path("/admin-account")
    @Consumes({"application/json", "application/xml"})
    @Produces({"application/json", "application/xml"})
    AccountResponse openAccountAdmin(@RequestBody(description = "Account to be saved", required = true, content = {@Content(mediaType = "application/json", schema = @Schema(implementation = OpenAccountRequest.class)), @Content(mediaType = "application/xml", schema = @Schema(implementation = OpenAccountRequest.class))}) OpenAccountRequest openAccountRequest);

    /**
     * Open User Account
     *
     * @param openAccountRequest
     * @return
     */
    @Operation(operationId = "openAccountUser", description = "openAccountUser",
            responses = {
                @ApiResponse(responseCode = "201", description = "Save User Account",
                    content = {
                            @Content(mediaType = "application/json", schema = @Schema(implementation = AccountResponse.class)),
                            @Content(mediaType = "application/xml", schema = @Schema(implementation =  AccountResponse.class))
                    },
                    headers = {
                            @Header(name = "Location", description = "URL of the created resource", schema = @Schema(type = "string"))
                    }),
                @ApiResponse(responseCode = "500", description = "Account whit this Id not found", content = {
                        @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                        @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                })
            }
    )
    @POST
    @Path("/user-account")
    @Consumes({"application/json", "application/xml"})
    @Produces({"application/json", "application/xml"})
    AccountResponse openAccountUser(@RequestBody(description = "Account to be saved", required = true, content = {@Content(mediaType = "application/json", schema = @Schema(implementation = OpenAccountRequest.class)), @Content(mediaType = "application/xml", schema = @Schema(implementation = OpenAccountRequest.class))}) OpenAccountRequest openAccountRequest);

    /**
     * Get Account
     *
     * @param paramLong
     * @return
     */
    @Operation(
            operationId = "getAccount",
            description = "Get Account",
            responses = {
                    @ApiResponse(
                            responseCode = "200",
                            description = "Account found",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = AccountResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = AccountResponse.class))
                            }),
                    @ApiResponse(responseCode = "404", description = "Account not found",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                        }
                    )
            })
    @GET
    @Path("/{id}")
    @Consumes({"application/json", "application/xml"})
    @Produces({"application/json", "application/xml"})
    Response getAccount(@PathParam("id") long paramLong);

    /**
     * Get Accounts
     *
     * @param getAccountsRequest
     */
    @Operation(
            operationId = "getAccounts",
            description = "Get Accounts",
            responses = {
                    @ApiResponse(
                            responseCode = "200",
                            description = "Accounts found",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = AccountPaginatedResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = AccountPaginatedResponse.class))
                            }),
                    @ApiResponse(responseCode = "404", description = "Accounts not found",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                            }
                    )
            }
    )
    @POST
    @Consumes({"application/json", "application/xml"})
    @Produces({"application/json", "application/xml"})
    @Path("/accounts")
    AccountPaginatedResponse getAccounts(@RequestBody(
            description = "GetAccountsRequest",
            required = true,
            content = {
                    @Content(mediaType = "application/json", schema = @Schema(implementation = GetAccountsRequest.class)),
                    @Content(mediaType = "application/xml", schema = @Schema(implementation = GetAccountsRequest.class))
            }
    ) GetAccountsRequest getAccountsRequest);

    /**
     * Promote User to Admin
     *
     * @param accountId
     */
    @Operation(
            operationId = "promoteAccount",
            description = "Promote Account to Admin",
            responses = {
                    @ApiResponse(
                            responseCode = "204",
                            description = "User promoted",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = Response.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = Response.class))
                            }),
                    @ApiResponse(responseCode = "404", description = "Account not found",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                            }
                    )
            }
    )
    @PATCH
    @Consumes({"application/json", "application/xml"})
    @Produces({"application/json", "application/xml"})
    @Path("/{accountId}/promote")
    Response promoteAccount(@PathParam("accountId") long accountId);

    /**
     * Demote User to Admin
     *
     * @param accountId
     */
    @Operation(
            operationId = "demoteAccount",
            description = "Demote Account to User",
            responses = {
                    @ApiResponse(
                            responseCode = "204",
                            description = "User demoted",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = Response.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = Response.class))
                            }),
                    @ApiResponse(responseCode = "404", description = "Account not found",
                            content = {
                                    @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class)),
                                    @Content(mediaType = "application/xml", schema = @Schema(implementation = ErrorResponse.class))
                            }
                    )
            }
    )
    @PATCH
    @Consumes({"application/json", "application/xml"})
    @Produces({"application/json", "application/xml"})
    @Path("/{accountId}/demote")
    Response demoteAccount(@PathParam("accountId") long accountId);
}
